src/samVisualize/graphGeneration/graph.ts

```diff
@@ -0,0 +1,118 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * Denotes a Link between two Nodes in the context of a GraphObject
+ */
+type Link = {
+    source: string
+    target: string
+    type?: string
+}
+
+/**
+ * Denotes a Node in the context of a GraphObject
+ */
+type Node = {
+    name: string
+    // Takes the form <a>::<b>::<c> eg. 'AWS::Serverless::Function'
+    type?: string
+}
+
+/**
+ * Denotes a graph derived from a Cloudformation template.
+ * Resources in the template are represented by Node objects.
+ * Conntections between resources are represented by Link objects.
+ */
+type GraphObject = {
+    nodes: Array<Node>
+    links: Array<Link>
+}
+
+class Graph {
+    /**
+     * Stores the internal representation of the graph
+     */
+    private graph: GraphObject
+
+    // Note: https://262.ecma-international.org/6.0/#sec-set-objects
+
+    // Set objects must be implemented using either hash tables or other mechanisms that, on average,
+    // provide access times that are sublinear on the number of elements in the collection
+
+    /**
+     * Stores which nodes exist in the graph
+     * Node names are used as keys.
+     * Exists to avoid iteration over the graph.nodes list to check for duplicates
+     */
+    private nodeSet: Set<string>
+
+    /**
+     * Stores which links exist in the graph
+     * The concatenation of the Link source, target, and type strings are used as keys because
+     * neither source, target, nor type properties are individually unique.
+     * The combination of all three properties determines a unique Link.
+     * Exists to avoid iteration over the graph.links list to check for duplicates
+     */
+    private linkSet: Set<string>
+
+    constructor() {
```

Comment:
Nits:
* Add `public` to all public classes (it's not necessary but we like it being explicit)
* Are you ever going to create a new `graph`, `nodeSet`, or `linkSet` within the lifecycle of the object? If not, make them all `private readonly`; you can still modify the values in the objects but the objects themselves will be guarded against being rewritten.

Reply:
No problem!

Reply:
Actually, I don't think` public` can be added to this class. 
![image](https://user-images.githubusercontent.com/45087499/125982142-0c9140ec-76ec-42f3-8858-7a53ab2540ec.png)
I couldn't find any other examples of a typescript `public class` in the source

Reply:
Whoops, I meant to the class fields and functions. Add `export` to the class declaration :)

Reply:
Oh, I see. Will do!